cluster/aks/composition.yaml

```diff
@@ -0,0 +1,68 @@
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+metadata:
+  name: xaks.azure.platformref.crossplane.io
+  labels:
+    provider: azure
+spec:
+  writeConnectionSecretsToNamespace: upbound-system
+  compositeTypeRef:
+    apiVersion: azure.platformref.crossplane.io/v1alpha1
+    kind: XAKS
+  resources:
+    - base:
+        apiVersion: compute.azure.crossplane.io/v1alpha3
+        kind: AKSCluster
+        spec:
+          dnsNamePrefix: platform-ref-azure-cluster
+          location: West US 2
+          version: "1.21.2"
+      patches:
+        - fromFieldPath: metadata.annotations[crossplane.io/external-name]
+          toFieldPath: metadata.annotations[crossplane.io/external-name]
+        - fromFieldPath: spec.id
+          toFieldPath: spec.resourceGroupNameSelector.matchLabels[azure.platformref.crossplane.io/app]
+        - fromFieldPath: spec.id
+          toFieldPath: spec.vnetSubnetIDSelector.matchLabels[azure.platformref.crossplane.io/app]
+        - fromFieldPath: spec.parameters.nodes.size
+          toFieldPath: spec.nodeVMSize
+          transforms:
+            - type: map
+              map:
+                small: Standard_B2s
+                medium: Standard_B4ms
+                large: Standard_B8ms
+        - fromFieldPath: "spec.parameters.nodes.count"
+          toFieldPath: "spec.nodeCount"
+        - fromFieldPath: spec.writeConnectionSecretToRef.namespace
+          toFieldPath: spec.writeConnectionSecretToRef.namespace
+        - fromFieldPath: spec.writeConnectionSecretToRef.name
+          toFieldPath: spec.writeConnectionSecretToRef.name
+          transforms:
+            - type: string
+              string:
+                fmt: "%s-aks"
+      connectionDetails:
+        - fromConnectionSecretKey: kubeconfig
+    - base:
+        apiVersion: helm.crossplane.io/v1beta1
+        kind: ProviderConfig
+        spec:
+          credentials:
+            source: Secret
+            secretRef:
+              namespace: upbound-systen
```

Comment:
i think we want to avoid hard coding namespaces into these resources, see the platform-ref-aws example at https://github.com/upbound/platform-ref-aws/blob/main/cluster/eks/composition.yaml#L196

And previous fixes for your reference in https://github.com/upbound/platform-ref-aws/pull/42 and https://github.com/upbound/platform-ref-aws/pull/44/files

Reply:
Definitely something I neglected to remove. Makes sense.

Reply:
ok, looking at it again, and. the issues linked here, I'm not sure I understand fully. I removed namespace from azure-default-provider.yaml, removed the echo-eof creation and replaced with kubectl apply -n upbound-system of the examples file.

Will leave this thread open until we confirm this is what you are suggesting.

Reply:
Ah, yes let me try to clarify further.  It's totally fine to include the namespace in the top level example yamls that the user will consume directly. They can edit those if they want to target another namespace, they have access to do that easily, so they aren't really "hard coded".

What we should really be avoiding is actually the hard coding of namespaces **within** the compositions themselves.  Those get embedded into the package at build time and then the user has no control over those namespaces at run time.  If they are using a different environment, say with `crossplane-system` instead of `upbound-system`, then the hard coded composition namespaces would not match what their environment needs.

Therefore, within the compositions, we should always be patching in the namespace from the XR, so it can dynamically match the user's environment instead of being hard coded to what was written at composition authoring time.

Note, after further inspection, I think you are doing this down below on line 59-60.  Your'e taking the XR's `spec.writeConnectionSecretToRef.namespace` field and patching it at runtime into this namespace value.  I think I was thrown off by there being a default value at all here - I think this would work, the patch will just overwrite the default value at runtime.

I'd go ahead and remove the default value period to minimize confusion - especially because there appears to be a typo in it, `upbound-systen` with a trailing `n` :)

Reply:
And I may have mislead you a bit @natereid72 - looking again at platform-ref-aws as a reference, it looks like we've determined there that the place where hard coding a namespace is not OK is only within the resources of a composition. I do see some instances of setting the namespace at the top level composition spec, e.g.: https://github.com/upbound/platform-ref-aws/blob/main/cluster/composition.yaml#L6, but not down inside the resources of the composition.

Do try to test out your change tomorrow and see if you run into anything. If the namespace is needed at the top level composition, then we can put it back, while keeping namespaces removed from the composed resources.

Perhaps the namespace at the top level composition serves as a default for where to write the connection secrets to, and the UI will override that if it needs? There's still a bit of ambiguity here on how this all works, sorry about that 😊

Reply:
I think I have your gist now, was missing that this was for composition and not provider config. removed all hard coded default composition write secret to namespace, added patches. Not sure if it broke something, need to clean up azure objects to test. Will mark as resolved for now.